### PR TITLE
Add github actions ci using industrial_ci

### PR DESCRIPTION
**Pull request by tylerjw**
Signed-off-by: Tyler Weaver <tylerjw@gmail.com>

### CODE REVIEW

**Comment by rhaschke** [comment]
Cool, `industrial_ci` migrated to github actions!? However, it's not yet triggered, is it?

**Comment by tylerjw** [comment]
> Cool, `industrial_ci` migrated to github actions!? However, it's not yet triggered, is it?

yup, I've used it on a few client projects and it works quite well.

I'm not sure why it didn't trigger... looking into that.

**Comment by tylerjw** [comment]
I made a PR within my fork here and the check triggered: https://github.com/tylerjw/warehouse_ros/pull/1

Maybe this is some sort of setting in the repo settings that's preventing it from working here.  @rhaschke I don't have access to the repo settings for this repo, can you see if there is something in them that would stop this from triggering actions?

**Comment by tylerjw** [comment]
This is what the settings page looks like on my fork.  Perhaps there is something else in the actual one that is preventing this?

**Comment by tylerjw** [comment]
Running the actions job on my fork seems to work for `foxy`.  Perhaps unsurprisingly it fails for `rolling` but I just removed that from the test as that seems out of the scope of this PR.  I'm starting with converting this for the ros2 branch, but there is no reason we couldn't use it for the ros1 branches also.

The only thing this doesn't do that travis does is the clang_format check.  clang_format can be added through ament_lint (if we build ament_lint from source until they release again).  for ros1 it might be possible to also use `ament_lint` for this also, I need to do more experimenting.  There is also the option of just copying the clang_format test from the `moveit_ci` code into a standalone test and running it using one of the script BEFORE/AFTER hooks (if a script returns non-zero in one of these it'll fail the job).

In the interest of migrating our smaller packages off travis incrementally I suggest we fix whatever is in the settings preventing this from running, then merge this and use this package to work out the details of clang_format testing.  Once we have that working we can make another pr to remove travis.  Then, hopefully this can be a good template for our other smaller repos that are on travis.  I still think circleci will be a better option for the moveit/moveit2 repos themselves.

**Comment by tylerjw** [comment]
One really awesome thing is that for these smaller packages, GitHub Actions is really fast.  See the run here on my fork: https://github.com/tylerjw/warehouse_ros/pull/1/checks?check_run_id=2133045095

It took 3min to run ci on this package.  Admittedly all it is doing its doing is doing a build as this package has no tests, but comparing that to the day long backlog in travis... this is a really nice improvement.

**Comment by rhaschke** [comment]
The `Actions` settings look exactly the same here. Maybe it requires the workflow to be configured once in the main repo first (before allowing for PR checks)? If it works for you, let's merge and try. Shouldn't we disable Travis as well?

**Comment by tylerjw** [comment]
> Shouldn't we disable Travis as well?

I think we should wait until after I get clang-format running via ament_lint or a standalone script.  I could be convinced to disable travis though for the ros2 branch and work on the clang-format stuff after this is merged.

**Comment by rhaschke** [comment]
For clang-format, I just got [pre-commit.com](http://pre-commit.com) running [for MTC](https://github.com/ros-planning/moveit_task_constructor/commit/5de9ce7203fedb8cf1af94aa56add763bb0f353d). The cool thing about this is: You can configure a git pre-commit hook locally, that runs exactly the same checks before each commit! If configured, it's not possible to commit badly formatted files anymore.

**Comment by tylerjw** [comment]
> For clang-format, I just got [pre-commit.com](http://pre-commit.com) running [for MTC](https://github.com/ros-planning/moveit_task_constructor/commit/5de9ce7203fedb8cf1af94aa56add763bb0f353d).

That is really nice.  I'll look into using that myself.  Is that something that applies to running in CI or just something we can ask people to setup in their own environments?

:facepalm: I should have looked at your link.

**Comment by tylerjw** [comment]
Ok, I added those and removed the travis script.  It is working on my fork: https://github.com/tylerjw/warehouse_ros/pull/1/checks

File: .clang-format [hunk 1]
```diff
@@ -51,16 +50,15 @@ SpaceAfterCStyleCast: false
...
+  AfterControlStatement: 'true'
+  AfterEnum : 'true'
+  AfterFunction : 'true'
+  AfterNamespace : 'true'
+  AfterStruct : 'true'
+  AfterUnion : 'true'
+  BeforeCatch : 'true'
+  BeforeElse : 'true'
+  IndentBraces : 'false'
 ...
```
Comment (tylerjw) [inline .clang-format:64]: these changes were founded by the yaml checker

File: .gitignore [hunk 2]
```diff
@@ -1,2 +1,2 @@
 *~
-*.vscode
\ No newline at end of file
+*.vscode
```
Comment (tylerjw) [inline .gitignore:2]: fixed missing eof

**Comment by rhaschke** [comment]
> > For clang-format, I just got [pre-commit.com](http://pre-commit.com) running [for MTC](https://github.com/ros-planning/moveit_task_constructor/commit/5de9ce7203fedb8cf1af94aa56add763bb0f353d).
> 
> That is really nice. I'll look into using that myself. Is that something that applies to running in CI or just something we can ask people to setup in their own environments?

I guess you already figured out that this is working for both, locally and in CI. Sorry, I was offline for a while.

**Review by rhaschke — APPROVED** [review]

**Comment by rhaschke** [comment]
I just noticed that caching still has some issues:
```
/usr/bin/tar --use-compress-program zstd -d -xf /home/runner/work/_temp/e30e12ab-0888-4305-a8dd-c80030eb1db5/cache.tzst -P -C /home/runner/work/warehouse_ros/warehouse_ros
/usr/bin/tar: ../../../../../github: Cannot mkdir: Permission denied
```
```
Post job cleanup.
Unable to reserve cache with key ccache-foxy-main, another job may be creating this cache.
```

**Comment by tylerjw** [comment]
merged and it passed: https://github.com/ros-planning/warehouse_ros/actions

:tada:

**Comment by tylerjw** [comment]
> I just noticed that caching still has some issues:
> 
> ```
> /usr/bin/tar --use-compress-program zstd -d -xf /home/runner/work/_temp/e30e12ab-0888-4305-a8dd-c80030eb1db5/cache.tzst -P -C /home/runner/work/warehouse_ros/warehouse_ros
> /usr/bin/tar: ../../../../../github: Cannot mkdir: Permission denied
> ```
> 
> ```
> Post job cleanup.
> Unable to reserve cache with key ccache-foxy-main, another job may be creating this cache.
> ```

where did you see that?  I can't find it in the logs.

**Comment by rhaschke** [comment]
I have seen that in the logs of the link you provided: https://github.com/tylerjw/warehouse_ros/pull/1/checks.
However, here, everything looks fine. Maybe, the cache ID should include the PR-ID?

**Comment by tylerjw** [comment]
Oh, weird.  I hadn't seen that.  I basically copied this config from here: https://github.com/ros-industrial/industrial_ci/blob/master/doc/industrial_ci_action.yml

**Comment by tylerjw** [comment]
I think if we included the PR-ID the cache wouldn't be nearly as useful as it is a ccache cache and each PR would re-create a new cache.

**Comment by rhaschke** [comment]
Travis used a hierachical cache-resolving strategy: 
1. Use the cache (if available) of the PR
2. Use the cache (if available) of the branch targeted

I'm curious how github actions handles caches.

**Comment by tylerjw** [comment]
[from the official documentation](https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#matching-a-cache-key)

```You can provide a list of restore keys to use when there is a cache miss on key. You can create multiple restore keys ordered from the most specific to least specific. The cache action searches for restore-keys in sequential order. When a key doesn't match directly, the action searches for keys prefixed with the restore key. If there are multiple partial matches for a restore key, the action returns the most recently created cache.```

I'll make a PR for this.

**Pull request merged**